src/models/StyledComponent.js

```diff
@@ -119,22 +153,52 @@ export default (ComponentStyle: Function, constructWithOptions: Function) => {
       }
     }
 
-    componentWillMount() {
+    injectStyles() {
+      const { componentStyle } = this.constructor
+      const styleSheet = this.context[CONTEXT_KEY] || StyleSheet.master
+      componentStyle.injectStyles(this.state.generatedClassName, styleSheet)
+    }
+
+    getInitialState() {
       const { componentStyle } = this.constructor
       const styledContext = this.context[CHANNEL_NEXT]
 
-      // If this is a staticaly-styled component, we don't need to the theme
+      // If this is a staticaly-styled component, we don't need the theme
       // to generate or build styles.
       if (componentStyle.isStatic) {
-        const generatedClassName = this.generateAndInjectStyles(
+        const generatedClassName = this.generateClassName(
           STATIC_EXECUTION_CONTEXT,
           this.props
         )
-        this.setState({ generatedClassName })
-        // If there is a theme in the context, subscribe to the event emitter. This
-        // is necessary due to pure components blocking context updates, this circumvents
-        // that by updating when an event is emitted
+        return { generatedClassName }
+        // If there is a theme in the context, read it
       } else if (styledContext !== undefined) {
+        const { getTheme } = styledContext
+        const theme = determineTheme(
+          this.props,
+          getTheme(),
+          this.constructor.defaultProps
+        )
+        const generatedClassName = this.generateClassName(theme, this.props)
+        return { theme, generatedClassName }
+      } else {
+        // eslint-disable-next-line react/prop-types
+        const theme = this.props.theme || {}
+        const generatedClassName = this.generateClassName(theme, this.props)
+        return { theme, generatedClassName }
+      }
+    }
+
+    componentDidMount() {
+      const { componentStyle } = this.constructor
+      const styledContext = this.context[CHANNEL_NEXT]
+
+      this.injectStyles()
+
+      // If there is a theme in the context, subscribe to the event emitter. This
+      // is necessary due to pure components blocking context updates, this circumvents
+      // that by updating when an event is emitted
+      if (!componentStyle.isStatic && styledContext !== undefined) {
         const { subscribe } = styledContext
         this.unsubscribeId = subscribe(nextTheme => {
```

Comment:
Theme may change between initial read (in constructor) and mount, so I think the subscribtion logic can be kept the same. We can add `sCU` or extend `PureComponent` to prevent unneeded rerenders. This would make sense anyway, as right now context theme changes trigger a rerender even if they're overriden by explicit props

See also: https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html#adding-event-listeners-or-subscriptions

Reply:
Good call 👍 let’s try to avoid unnecessary rerenders by checking whether the theme hasn’t changed. (sCU / PureComponent would be the wrong tools though since it’s introduce a huge slowdown in components like with SC that are spread throughout an app. Also I’d love not to introduce any unexpected behaviour) It’d be nice if we can just introduce a small check in the subscription that’d compare whether it’s execution is necessary.

tl;dr themes are expected to be immutable, (shallow) props not necessarily

Reply:
Should we rather check that `generatedClassName` has changed then? There's a chance that theme has changed in a way that doesn't affect this particular component

Reply:
Oh, I don't think it's possible: `prevState.theme` should be up to date when `gDSFP` is called next time

Reply:
Or we can intoduce a `shouldUpdate` field on state which could be the only thing checked in `sCU`. Then `gDSFP` would always set it to `true`, and subscription would set it to `false` if `generatedClassName` hasn't changed

Reply:
@Hypnosphi I don't think a `shouldUpdate` field makes sense, sorry 😅 We can trust that the theme is pure, since that's an assumption we already made a while ago (I believe). If not I think we should make it now either way.

Regenerating styles, reinjection, etc, that's something we'd like to avoid together with an unnecessary rerender. The former is theoretically often more expensive even than the latter

Reply:
Or maybe even simpler:

```
shouldComponentUpdate(nextProps, nextState) {
  // React passes a new props object on each external rerender
  return nextProps !== prevProps || nextState.generatedClassName !== this.state.generatedClassName
}
```

The first part of this check is extra cheap, and it guarantees that component keeps being rerendered when its parent rerenders

Reply:
> Regenerating styles, reinjection, etc, that's something we'd like to avoid together with an unnecessary rerender. The former is theoretically often more expensive even than the latter

OK, got your point

Reply:
Oh, I think we shouldn't try to guess how the theme has changed. At that point it's pretty cheap to say that the theme has changed and perform the style generation & injection again. Then it's up to the stylesheet to deduplicate and running the render on the classname should be cheap enough so that we don't have to decide whether to do it or not.

Theme changes should be rare anyway 💯